### PR TITLE
Poc idm billing service

### PR DESCRIPTION
**Pull request by NCorreia100**
## Link to JIRA:
[ID-486](https://clever.atlassian.net/browse/ID-486)

## Overview:
Bare bones microservice that authorizes payments issued via stripe
If you made any changes to swagger.yml:
- [x] Update swagger.yml version
- [x] Run "make generate"

## Rollback (in the event of a problem)
- [ ] Rollback via dapple OR `ark rollback -e production idm-billing-service`?
- [ ] Anything else?

## New Repo Setup
- [ ] Set up Slack notifications for this app for your team https://clever.atlassian.net/wiki/spaces/ENG/pages/888897571/GitHub+assignments
- [ ] Follow the instructions at go/private-deps
- [ ] Tune your alarms if you'd like anything other than our default recommendations https://clever.atlassian.net/wiki/spaces/~620990898/pages/904036784/Alarm+Best+Practices
- [ ] Adjust the cpu and max_mem values based on the needs of your application as the containers are killed if these values are exceeded
- [ ] If this app does not need to be canaried, you can disable that in the `deploy_config` section of the launch config
- [ ] If this app should be multi region in sso pods adjust the pod_config https://clever.atlassian.net/wiki/spaces/ENG/pages/370147335/launch.yml#launch.yml-PodConfig
- [x] Run "make install_deps" and commit the go.sum file
- [ ] Delete this section from the PR template

Screen Shot 2021-05-21 at 1 16 52 PM
Screen Shot 2021-05-21 at 3 10 35 PM
Screen Shot 2021-05-21 at 3 16 02 PM

### CODE REVIEW

**Review by johnhuangclever — APPROVED** [review]

File: swagger.yml [hunk 1]
```diff
@@ -37,39 +37,33 @@ paths:
...
-      responses:
-        200:
-          description: OK response
     post:
       summary: Validates the payment made with Stripe
       operationId: authorizePayment
       parameters: 
-        - name: authorizePayment
-          in: body
-          required: true
```
[the remark this answers is not in the document]
Comment (NCorreia100) [reply swagger.yml:52]: Re-added.

File: swagger.yml [hunk 2]
```diff
@@ -83,9 +77,11 @@ definitions:
       token:
         type: string
       amount:
-        type: number
-      adminID:
+        type: integer
+      email:
         type: string
+      districtID:
```
[the remark this answers is not in the document]
Comment (NCorreia100) [reply swagger.yml:83]: Not absolutely necessary to issue a charge. We only need them to identify the charges & customers from the dashboard.

File: main.go [hunk 3]
```diff
@@ -36,25 +38,74 @@ func (mc MyController) HealthCheck(ctx context.Context) error {
...
+	completedCharge, err := charge.New(&stripe.ChargeParams{
+		Amount:   stripe.Int64(*i.Amount),
+		Currency: stripe.String(string(stripe.CurrencyUSD)),
+		Customer: stripe.String(newCustomer.ID),
+
+		// ReceiptEmail: stripe.String(i.input.email), //might be useful
+		//ID: stripe.String(districtID) //might be useful
+		//Metadata:  //might be useful
+	})
+	fmt.Println(completedCharge)
```
[the remark this answers is not in the document]
Comment (NCorreia100) [reply main.go:87]: Done

File: main.go [hunk 4]
```diff
@@ -36,25 +38,74 @@ func (mc MyController) HealthCheck(ctx context.Context) error {
...
+
+	if err != nil {
+		return nil, models.PreconditionFailed{
+			Message: fmt.Sprintf("Payment failed to process", err.Error()),
+			//note that even if payment fails a new customer is still issued on stripe
+		}
+	}
+
+	// fmt.Printf("request body %v", json.Marshal(i))
```
[the remark this answers is not in the document]
Comment (NCorreia100) [reply main.go:96]: Done

File: swagger.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (johnhuangclever) [reply swagger.yml:83]: So from what I understand, they are not required for the stripe API, but are they required for the Clever use case? This struct schema should represent the Clever use case if we can

**Pull request merged**